docs: Clean up migration guide for v9 (#15287)
getsentry/sentry-javascript: docs/migration/v8-to-v9.md

### TypeScript Version Policy

In preparation for v2 of the OpenTelemetry SDK, which will raise the minimum required TypeScript version, the minimum required TypeScript version is increased to version `5.0.4`.

Additionally, like the OpenTelemetry SDK, the Sentry JavaScript SDK will follow [DefinitelyType's version support policy](https://github.com/DefinitelyTyped/DefinitelyTyped#support-window) which has a support time frame of 2 years for any released version of TypeScript.

Older Typescript versions _may_ continue to be compatible, but no guarantees apply.

## 2. Behavior Changes

### `@sentry/core` / All SDKs

- Passing `undefined` as a `tracesSampleRate` option value will now be treated the same as if the option was not defined at all.
  In previous versions, it was checked whether `tracesSampleRate` or `traceSampler` are _defined_ at all, in `Sentry.init()` to determine whether tracing should be enabled.
  Passing `undefined` to `tracesSampleRate` previously would enable tracing without the performance monitoring aspect (=tracing being enabled, though no spans would be started).

>   If you want to preserve having tracing enabled without creating spans, set `tracesSampleRate: 0`:

Reviewer: m: Setting tracesSampleRate:0 doesn't enable TwP. Not setting the rate at all or setting it to `undefined` should enable TwP. https://develop.sentry.dev/sdk/telemetry/traces/tracing-without-performance/#configuration

setting a rate of 0 also forces downstream SDKs to always sample negatively.

Author: https://github.com/getsentry/sentry-javascript/pull/15287/commits/ceb88b0f4593a0c93669c5c5c342fca55fd55b24

  ```diff
  Sentry.init({
  -  tracesSampleRate: undefined,
  +  tracesSampleRate: 0,
  });
  ```

- Dropping spans in the `beforeSendSpan` hook is no longer possible.
  This means you can no longer return `null` from the `beforeSendSpan` hook.
  This hook is intended to be used to add additional data to spans or remove unwanted attributes (for example for PII stripping).
  To control which spans are recorded, we recommend configuring [integrations](https://docs.sentry.io/platforms/javascript/configuration/integrations/) instead.

- The `beforeSendSpan` hook now receives the root span as well as the child spans.
  We recommend checking your `beforeSendSpan` to account for this change.

> - If you use the optional `captureConsoleIntegration` and set `attachStackTrace: true` in your `Sentry.init` call, console messages will no longer be marked as unhandled (i.e. `handled: false`) but as handled (i.e. `handled: true`).

Author: I am gonna move it down within the section but not move it across sections as it still thematically fits behavior changes.

     }
     ```

- When `skipOpenTelemetrySetup: true` is configured, `httpIntegration({ spans: false })` will be configured by default.
  You no longer have to specify this manually.
  With this change, no spans are emitted once `skipOpenTelemetrySetup: true` is configured, without any further configuration being needed.

### `@sentry/browser`

> - The SDK no longer instructs the Sentry backend to automatically infer IP addresses by default.

Author: 5c4621131d914dbf9cc2e3d73bbb7e9e348544d3

  This means that places where you previously saw IP addresses in Sentry may now be grouped to anonymous users.
  Set the `sendDefaultPii` option in `Sentry.init()` to true to instruct the Sentry backend to infer IP addresses.
- The `captureUserFeedback` method has been removed.

>   Use the `captureFeedback` method instead and update the `comments` field to `message`.

Reviewer 2: ```suggestion
  Use the `captureFeedback` method instead and set the `message` property instead of `comments`.
```

  In the `onUnmount` lifecycle method, the `componentStack` field is now typed as `string | null`. The `componentStack` is `null` when no error has been thrown at time of unmount.

## 3. Package Removals

The `@sentry/utils` package will no longer be published.

The `@sentry/types` package will continue to be published, however, it is deprecated and its API will not be extended.
It will not be published as part of future major versions.

All exports and APIs of `@sentry/utils` and `@sentry/types` (except for the ones that are explicitly called out in this migration guide to be removed) have been moved into the `@sentry/core` package.

## 4. Removed APIs

### `@sentry/core` / All SDKs

- **The metrics API has been removed from the SDK.**

  The Sentry metrics beta has ended and the metrics API has been removed from the SDK. Learn more in the Sentry [help center docs](https://sentry.zendesk.com/hc/en-us/articles/26369339769883-Metrics-Beta-Ended-on-October-7th).

- The `enableTracing` option was removed.

>   In v9, to emulate `enableTracing: true`, set `tracesSampleRate: 1`.
>   To emulate `enableTracing: false`, remove the `tracesSampleRate` and `tracesSampler` options (if configured).

Reviewer 2: ```suggestion
  Instead, set `tracesSampleRate: 1` or `tracesSampleRate: 0`.
```

This is good enough IMHO, as even before `tracesSampler` took precedence over `enableTracing`, so that does not change!

  });
  ```

> - The `transactionContext` property on the `samplingContext` argument passed to the `tracesSampler` and `profilesSampler` options has been removed. All object attributes are available in the top-level of `samplingContext`.

Reviewer 2: For me this change falls into the category: "we should highlight this somehow up top", as it cannot be caught nicely by linters (I think) and is very possible to actually affect users at runtime. 🤔

Author: True, and we also should add a code example I realized!

- The `debugIntegration` has been removed. To log outgoing events, use [Hook Options](https://docs.sentry.io/platforms/javascript/configuration/options/#hooks) (`beforeSend`, `beforeSendTransaction`, ...).

- The `sessionTimingIntegration` has been removed. To capture session durations alongside events, use [Context](https://docs.sentry.io/platforms/javascript/enriching-events/context/) (`Sentry.setContext()`).

### Server-side SDKs (`@sentry/node` and all dependents)

- The `addOpenTelemetryInstrumentation` method was removed.
  Use the `openTelemetryInstrumentations` option in `Sentry.init()` or your custom Sentry Client instead.

- `registerEsmLoaderHooks` now only accepts `true | false | undefined`.
  The SDK will default to wrapping modules that are used as part of OpenTelemetry Instrumentation.

- The `nestIntegration` was removed.
  Use the NestJS SDK (`@sentry/nestjs`) instead.

- The `setupNestErrorHandler` was removed.
  Use the NestJS SDK (`@sentry/nestjs`) instead.

### `@sentry/browser`

> - The `captureUserFeedback` method has been removed. Use the `captureFeedback` method instead and update the `comments` field to `message`.

Reviewer 2: this is _also_ mentioned above already, once is probably fine? 🤔

> - The `captureUserFeedback` method has been removed. Use the `captureFeedback` method instead and update the `comments` field to `message`.

Reviewer: super-l:
```suggestion
- The `captureUserFeedback` method was removed. Use the `captureFeedback` method instead and update the `comments` field to `message`.
```

  });
  ```

- The option `logErrors` in the `vueIntegration` has been removed. The Sentry Vue error handler will always propagate the error to a user-defined error handler or re-throw the error (which will log the error without modifying).

### `@sentry/nuxt`

- The `tracingOptions` option in `Sentry.init()` was removed in favor of passing the `vueIntegration()` to `Sentry.init({ integrations: [...] })` and setting `tracingOptions` there.

### `@sentry/vue` and `@sentry/nuxt`

- When component tracking is enabled, "update" spans are no longer created by default.

  Add an `"update"` item to the `tracingOptions.hooks` option via the `vueIntegration()` to restore this behavior.

  ```ts
  Sentry.init({
    integrations: [
      Sentry.vueIntegration({
        tracingOptions: {
          trackComponents: true,
          hooks: [
            'mount',
            'update', // add this line to re-enable update spans
            'unmount',
          ],
        },
      }),
    ],
  });
  ```

### `@sentry/remix`

- The `autoInstrumentRemix` option was removed.

>   The SDK now behaves as if the option were set to `true` as the only and default behavior.

Reviewer 2: ```suggestion
  The SDK now always behaves as if the option were set to `true`.
```
